### PR TITLE
Clean up comments in nestkernel

### PR DESCRIPTION
**Pull request by jessica-mitchell**
This PR aims to make the comments (doxygen and c++) more consistent and useful.
So far, only a few files have been touched because I would like some feedback as to whether I'm going in the right direction before proceeding. @jougs could you take a look?

Changes implemented:

* C++ comments single line (max 2 lines if sentence is long): // 
* C++ comments multile line: /* * * * */
* doxygen comments single line (max 2 lines if sentence is long): //!
* doxygen comments multiline: /** * * * * */
* doxygen comments side: //!<
* avoid doxygen comments in .cpp files, move to header file 
* avoid duplicating code in comments; only include code(parameters, functions etc) that also have additional context,  remove redundancies
* include the variable name in functions in header file to match cpp file.

### CODE REVIEW

File: nestkernel/conn_builder.h
Comment (jougs) [inline nestkernel/conn_builder.h:124]: This should fix the build problems:

```suggestion
  bool change_connected_synaptic_elements( index snode_id, index tnode_id, const int tid, int update );
```

**Review by jougs — COMMENTED** [review]
Thanks for making the chaos and invonsistencies visible. I'm reviewing in greater detail once the crying stops. For now, I have a couple of generic comments:

* I prefer the following doxygen style for functions and classes, even if the comment itself is only a single line:
  ```
  /**
   * ...
   */
  ```
  (I also think this is the predominant style presently used, so it will save you a lot of trouble down the line :smile:)
* I would use `//!` for long one-liners above variables and `//!<` for short ones behind them

Maybe we should set up a video session where we go through the [Doxygen documentation](https://www.doxygen.nl/manual/docblocks.html) and ask people what they (dis)like most?

File: nestkernel/clopath_archiving_node.cpp
Comment (jougs) [inline nestkernel/clopath_archiving_node.cpp:75]: Here's another one for fixing the build
```suggestion
  delayed_u_bars_idx_ = 0;
```

**Comment by jessica-mitchell** [comment]
> Thanks for making the chaos and invonsistencies visible. I'm reviewing in greater detail once the crying stops. For now, I have a couple of generic comments:
> 
>     * I prefer the following doxygen style for functions and classes, even if the comment itself is only a single line:
>       ```
>       /**
>        * ...
>        */
>       ```
Sure, I think you're right that it's more common
> 
>     * I would use `//!` for long one-liners above variables and `//!<` for short ones behind them

When you say above variables, can you give me an example of the syntax you'd expect this kind of comment? Where would not want this comment style? (it's just easier if I have an example to look up, because it's easy to get confused staring at slashes and askterisks all day ;) )
> 
> Maybe we should set up a video session where we go through the [Doxygen documentation](https://www.doxygen.nl/manual/docblocks.html) and ask people what they (dis)like most?

Yah sure, I think that could work.

**Comment by jougs** [comment]
Here are some examples that hopefully clarify my style preferences mentioned above:
```c++
/**
 * Set the minimal and maximal delay and override automatically determined values.
 */
void set_delay_extrema(delay min_delay, delay, max_delay();

//! The variable min_delay holds the value for the smallest transmission delay in the network in steps
delay min_delay;

delay max_delay //!< The largest transmission delay in the network (steps)
```

In real code, I would then actually prefer if similar variables were documented consistently, i.e. both `min_delay` and `max_delay` with either the long version, or with the short one, depending on the length of the comment.

File: doc/devdoc/cppcomments.rst
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:5]: This looks a bit confusing. I'd typeset it over multiple lines.

File: doc/devdoc/cppcomments.rst
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:15]: I would typeset this over multiple lines as well.

File: doc/devdoc/cppcomments.rst
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:22]: At the beginning of sentences, I'd capitalize in keeping with the doxygen website.
```suggestion
Doxygen comments single line (max 2 lines if sentence is long): //!
```

Should there be backticks around //!, and similar below?

File: doc/devdoc/cppcomments.rst
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:3]: Maybe add a link to doxygen documentation?

File: doc/devdoc/cppcomments.rst
Comment (heplesser) [inline doc/devdoc/cppcomments.rst:2]:
```suggestion
Rules for C++ code comments
===========================
```

File: nestkernel/archiving_node.h
Comment (heplesser) [inline nestkernel/archiving_node.h:54]: Trivial comments like this on constructor, copy constuctor or assignment operator provide no information. I would remove them completely.

File: nestkernel/archiving_node.h
Comment (heplesser) [inline nestkernel/archiving_node.h:63]: If I am not entirely mistaken, the same rules apply to doxygen comments as to Python docstrings: the first line should give a short summary of what a function does, followed by an empty line and then, if relevant, details. So here 
```c++
/**
 * Return the Kminus (synaptic trace) value at t (in ms). 
 * 
 * When the trace is ...
```

File: nestkernel/archiving_node.h
Comment (heplesser) [inline nestkernel/archiving_node.h:135]:
```suggestion
  /*
   * Number of incoming connections from stdp connectors.
   *
   * It is needed to determine if every incoming connection has
   * read the spike history for a given point in time.
   */
```

File: nestkernel/clopath_archiving_node.cpp
Comment (heplesser) [inline nestkernel/clopath_archiving_node.cpp:175]: I'd put an empty line above this comment.

**Review by heplesser — CHANGES_REQUESTED** [review]
@jessica-mitchell I added some comments to the guidelines as two examples and would like your and @jougs' feedback before commenting more.

**Comment by terhorstd** [comment]
@jessica-mitchell, do you have a status update on this one?

**Comment by jougs** [comment]
@jessica-mitchell: What is the status of this? Thanks!

File: doc/devdoc/cppcomments.rst
Comment (jessica-mitchell) [reply doc/devdoc/cppcomments.rst]: done

File: doc/devdoc/cppcomments.rst
Comment (jessica-mitchell) [reply doc/devdoc/cppcomments.rst]: yes

File: nestkernel/archiving_node.h
Comment (jessica-mitchell) [reply nestkernel/archiving_node.h]: Removed these, also removed destructor comments as well

File: nestkernel/archiving_node.h
Comment (jessica-mitchell) [reply nestkernel/archiving_node.h]: Added a new line after first line description in comments

File: doc/devdoc/cppcomments.rst
Comment (jessica-mitchell) [reply doc/devdoc/cppcomments.rst]: done

File: doc/htmldoc/developer_space/cppcomments.rst
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/cppcomments.rst:1]:
```suggestion
Guidelines for C++ code comments
```

File: nestkernel/conn_builder.h
Comment (jessica-mitchell) [inline nestkernel/conn_builder.h:356]:
```suggestion
```

File: doc/htmldoc/developer_space/cppcomments.rst
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/cppcomments.rst:12]:
```suggestion
Generate HTML from doxygen comments
```

File: doc/htmldoc/developer_space/cppcomments.rst
Comment (jessica-mitchell) [inline doc/htmldoc/developer_space/cppcomments.rst:13]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by jessica-mitchell** [comment]
@jougs @heplesser I think I addressed all the comments so far. In addition to removing the Constructor comments, I also removed Destructor.

I cleaned up the guidelines (doc/htmldoc/developer_space/cppcoments.rst) to be more clear, but I am unsure of how to describe when to use C++ comment style `/* */` versus doxygen style `/** */`.

File: nestkernel/connection.h
Comment (jougs) [inline nestkernel/connection.h:346]: Revert

File: nestkernel/connection_id.h
Comment (jougs) [inline nestkernel/connection_id.h:109]: Revert

**Review by jougs — COMMENTED** [review]
Here's a bunch of comments. Some of them are rather generic and should be applied in many places. Probably something like `git difftool --dir-diff upstream/master` could help with bulk reverting changes more easily than purely by hand. I will have another look once my comments are addressed.

File: nestkernel/common_properties_hom_w.h
Comment (jessica-mitchell) [reply nestkernel/common_properties_hom_w.h:37]: I would propose that a separate PR be made to deal with any of  these SLI documentation sections that needs to be converted into reStructured Text

File: nestkernel/conn_builder_conngen.h
Comment (jessica-mitchell) [reply nestkernel/conn_builder_conngen.h]: done

File: nestkernel/conn_builder_conngen.cpp
Comment (jessica-mitchell) [inline nestkernel/conn_builder_conngen.cpp:298]:
```suggestion
#endif /* ifdef HAVE_LIBNEUROSIM  */
```

File: nestkernel/deprecation_warning.cpp
Comment (jessica-mitchell) [inline nestkernel/deprecation_warning.cpp:61]:
```suggestion
#endif /* DEPRECATION_WARNING_IMPL_H  */
```

File: nestkernel/mpi_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:98]:
```suggestion
#else /* HAVE_MPI */
```

File: nestkernel/mpi_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:127]:
```suggestion
#else  /* #ifdef HAVE_MUSIC */
```

File: nestkernel/mpi_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:131]:
```suggestion
#endif /* #ifdef HAVE_MUSIC */
```

File: nestkernel/mpi_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:279]:
```suggestion
#else /* #ifdef HAVE_MPI */
```

File: nestkernel/mpi_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:981]:
```suggestion
#else /* #ifdef */
```

File: nestkernel/mpi_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.cpp:1093]:
```suggestion
#endif /* #ifdef HAVE_MPI  */
```

File: nestkernel/mpi_manager.h
Comment (jessica-mitchell) [inline nestkernel/mpi_manager.h:771]:
```suggestion
#endif /* HAVE_MPI  */
```

File: nestkernel/mpi_manager_impl.h
Comment (jessica-mitchell) [inline nestkernel/mpi_manager_impl.h:49]:
```suggestion
#else  /* #ifdef HAVE_MUSIC */
```

File: nestkernel/music_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/music_manager.cpp:100]:
```suggestion
#else /* #ifdef HAVE_MUSIC */
```

File: nestkernel/music_manager.cpp
Comment (jessica-mitchell) [inline nestkernel/music_manager.cpp:127]:
```suggestion
#else /* #ifdef HAVE_MUSIC */
```

File: nestkernel/common_properties_hom_w.h
Comment (jougs) [reply nestkernel/common_properties_hom_w.h]: I would not bother with any SLI things anymore at this point as these comments will anyway just be deleted in PyNEST NG together with all of SLI.

**Comment by jessica-mitchell** [comment]
@jougs the conflict is fixed

**Comment by jessica-mitchell** [comment]
@jougs ping!

**Comment by jougs** [comment]
@jessica-mitchell: [pong!](https://github.com/jessica-mitchell/nest-simulator/pull/19)

**Review by jougs — APPROVED** [review]
Many thanks for this heroic effort and your patience. Merging without second review due to discussion in several meetings and because this "only" touches documentation.

**Pull request merged**
